### ISSUE TITLE
Emulated 6 button controller (Genesis/Mega Drive)

### ISSUE THREAD

**Issue by niuus** [raz0red/wii-mednafen#64]
6 button controller doesn't seem to be entirely working on the Genesis/Mega Drive (6 buttons worked perfect on Turbografx 16, in case you're curious). My only test game was Street Fighter II Champion Edition, A B C work, while X Y Z do not. The controller is properly detected in-game when you open the game's Options menu.

Happening in the 2019 temporary build, and 0.2pre

**Comment by raz0red** [comment]
This should be resolved based on PR #71, and is available in the latest snapshot builds:

https://github.com/raz0red/wii-mednafen/actions

If you click on an individual build, you will see the status of the build. If the build was successful you will see an "Artifacts" portion in the upper-right. If you download the artifact named "dist", you will find the distribution .zip file within it.

The 6-button controls worked fine up until version 0.2-pre. In that version, I added extended controls support for NES. Unfortunately, those changes broke 6-button support for MD/Genesis.

At any rate, it should be working correctly now. It should also be noted that some games (Mortal Kombat, maybe more), require the game to be "RESET" if controls are changed (3 to 6 button, etc.) while the game is running. Most other games support selecting the controller type within a menu in the game, allowing for hot swapping of controller types.

**Comment by raz0red** [comment]
> This should be resolved based on PR #71, and is available in the latest snapshot builds:
> 
> https://github.com/raz0red/wii-mednafen/actions
> 
> If you click on an individual build, you will see the status of the build. If the build was successful you will see an "Artifacts" portion in the upper-right. If you download the artifact named "dist", you will find the distribution .zip file within it.
> 
> The 6-button controls worked fine up until version 0.2-pre. In that version, I added extended controls support for NES. Unfortunately, those changes broke 6-button support for MD/Genesis.
> 
> At any rate, it should be working correctly now. It should also be noted that some games (Mortal Kombat, maybe more), require the game to be "RESET" if controls are changed (3 to 6 button, etc.) while the game is running. Most other games support selecting the controller type within a menu in the game, allowing for hot swapping of controller types.

I just noticed that the "artifacts" tab is not available on GitHub unless you are logged in. However, if you look at the "Commit History" you will see a green check mark or red x depending on whether the build passed or failed. If you click on the check mark or x, it will take you to the automated build. From there, you can download the artifact in the upper right corner (if it was a passing build).

**Comment by niuus** [comment]
It seems to be partially fixed, raz0red. Did not test all the 6 buttons games that exist, but here's a list of the ones where the controller gets detected (as noted in their _Options_ menu), but does not register any input in-game:

- Comix Zone
- Eternal Champions
- Mortal Kombat 3
- Mortal Kombat II (*)
- Mortal Kombat (*)
- Shaq-Fu
- Streets of Rage 3
- Ultimate Mortal Kombat 3
- Zombies Ate My Neighbors

(*) means the game does not indicate anything special about the 6 button controller being plugged.
NOTE: i am always using a reset after changing from 3 to 6b.

Basically, the only games i discovered had working 6b inputs were _Contra Hard Corps_ and _Street Fighter II_

**Comment by raz0red** [comment]
Hi-

Thanks for doing a great job testing, I really appreciate it.

I just tested Zombies Ate My Neighbors and it appeared to be working as expected. I did the following:

1.) Loaded the game
2.) Went into game-specific settings and turned on 6 button (for both controllers).
3.) I Mapped the X, Y, and A, B buttons of the Genesis to the Wiimote (Using A,B,1,2).
4.) Saved the game-specific settings
5.) Re-loaded the game (didn't do a reset, but it should work as well, but maybe not).

At that point, I could use and switch weapons as well as using and switch items.

Also, there is a button on the genesis controller to swap between 3 and 6 button mode. Make sure you don't press that as it can cause things to not work correctly (because you would be in 3 button mode).

I was also able to get Mortal Kombat to work. I did the same procedure as above. Then, once I loaded the game, I went into "Options". There will be a new menu item "Extra controls". In that menu, you need to switch it to "6-button".

Please let me know if you are still having issues. I can go test these other games as well. Maybe there is something different in our configuration that is causing it to not work for you. I am using a Wiimote.

Thanks again for the great help testing!
